src/core/middlewares/util/util.go

```diff
@@ -0,0 +1,168 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"encoding/json"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/utils/clair"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/core/promgr"
+	"net/http"
+	"net/http/httptest"
+	"regexp"
+	"strings"
+)
+
+type contextKey string
+
+const (
+	manifestURLPattern = `^/v2/((?:[a-z0-9]+(?:[._-][a-z0-9]+)*/)+)manifests/([\w][\w.:-]{0,127})`
+	blobURLPattern     = `^/v2/((?:[a-z0-9]+(?:[._-][a-z0-9]+)*/)+)blobs/uploads/`
+	// ImageInfoCtxKey the context key for image information
+	ImageInfoCtxKey = contextKey("ImageInfo")
+	// TokenUsername ...
+	// TODO: temp solution, remove after vmware/harbor#2242 is resolved.
+	TokenUsername = "harbor-core"
+)
+
+// ImageInfo ...
+type ImageInfo struct {
+	Repository  string
+	Reference   string
+	ProjectName string
+	Digest      string
+}
+
+// JSONError wraps a concrete Code and Message, it's readable for docker deamon.
+type JSONError struct {
+	Code    string `json:"code,omitempty"`
+	Message string `json:"message,omitempty"`
+	Detail  string `json:"detail,omitempty"`
+}
+
+// MarshalError ...
+func MarshalError(code, msg string) string {
+	var tmpErrs struct {
+		Errors []JSONError `json:"errors,omitempty"`
+	}
+	tmpErrs.Errors = append(tmpErrs.Errors, JSONError{
+		Code:    code,
+		Message: msg,
+		Detail:  msg,
+	})
+	str, err := json.Marshal(tmpErrs)
+	if err != nil {
+		log.Debugf("failed to marshal json error, %v", err)
+		return msg
+	}
+	return string(str)
+}
+
+// MatchManifestURL ...
+func MatchManifestURL(req *http.Request) (bool, string, string) {
+	re := regexp.MustCompile(manifestURLPattern)
```

Comment:
Should use `Compile` instead, the `MustCompile` panics if the pattern cannot be parsed.